Blog: add posts, tags, versions for blogs

juhalehtonen opened the pull request:
Part of the blog support project.

Adds database schemas for blog posts, tags and versions. Versions are meant to be embedded into posts, hence no separate migration for them.

This PR does not use or alter the existing blog post schema, because the overarching idea is that we want to be able to deploy new blog-related features to production without affecting how the existing site functions until we are ready to switch over to a new system.

Happy to hear thoughts on proposed structure. A notable discussion point is the absence of a category field, as the combination of `working_group_id` and `blog_tags` is proposed as an alternative.

- [x] I acknowledge my contribution to the website does not assert comparative or superlative differences of one product, project, company or individual over another.

starbelly commented on lib/erlef/erlef/blog/post.ex:
```diff
@@ -0,0 +1,29 @@
+  schema "blog_posts" do
+    field(:body, :string)
+    field(:slug, :string)
+    field(:status, :string)
```
What do you think about making status an enum (ecto enum) and checking the status in the changeset vs on the db side? Rationale is 1. the application can use atoms for the status 2. Since performance isn't a concern, we can gain flexibility by only having to modify code in the case of adding or removing valid statuses?

juhalehtonen replied:
Great suggestion! Will definitely go with `Ecto.Enum` here, and drop the constraint-based approach. Now that I think of it, the definition of different statues is definitely a business-side requirement anyway, not something the DB should care this much about.

starbelly commented on lib/erlef/erlef/blog/post.ex:
```diff
@@ -0,0 +1,29 @@
+  def changeset(post, attrs) do
+    post
+    |> cast(attrs, [:title, :slug, :body, :status])
+    |> validate_required([:title, :slug, :body, :status])
```
Do you think `member_id` should be required?

juhalehtonen replied:
Yes, definitely! Thanks for spotting that.

starbelly commented:
@juhalehtonen This looks great so far! Added a comment/question. Also could we get some basic tests for the schemas?

juhalehtonen commented:
Hi @starbelly ! I've now addressed your previous comments, added a handful of tests for post and tag schemas/changesets, and reworked some parts of the code. It might be worth having another look through for anything else that might be poking out.

One particular point of interest is not casting a slug for post/tag in the changeset, but instead building one automatically within the changeset. I saw that something akin to this was already in place in `Erlef.Integrations.App`, so I gave it a go. However, I do not quite know what to think of this, and would appreciate hearing your thoughts :) 

What's bothering me about the approach I took here is that I see changesets to be about validating the data, whereas now I'm adding additional values to the resulting changeset. What I might prefer to do here is to drop the slugification from the changeset and add it to the (yet-to-come) blog context module that handles more of the post/tag creation business logic.

starbelly commented:
> What's bothering me about the approach I took here is that I see changesets to be about validating the data, whereas now I'm adding additional values to the resulting changeset. What I might prefer to do here is to drop the slugification from the changeset and add it to the (yet-to-come) blog context module that handles more of the post/tag creation business logic.

That makes sense to me. If we can keep the changeset function dumb, then we should.

starbelly commented on lib/erlef/blog/post.ex:
```diff
@@ -0,0 +1,37 @@
+    field(:body, :string)
+    field(:status, Ecto.Enum, values: [:draft, :published, :archived])
+
+    field(:member_id, :id)
```
I think both of the `:member_id` field and `:working_group_id` should be `belongs_to`. Maybe there was a reason not to go this route though? This if of course devoid of what we require.

starbelly commented on lib/erlef/blog/post.ex:
```diff
@@ -0,0 +1,37 @@
+  @moduledoc """
+  Schema for blog posts.
+  """
+  use Ecto.Schema
```
To note you can make use of `use Erlef.Schema` here which will : 

- use Ecto.Schema
- import Ecto.Changeset
- import Erlef.Schema itself for helpers 
- set primary key options (uuid) 
- set timestamp opts 

I'm not normally a huge fan of uuids as a default, but that's what we're using across the board and the data is so small that running into performance issues isn't a concern, to note.

starbelly commented:
Sorry @juhalehtonen we're going to close this in favor of a fresh start for the sake of continuity. Thank for your efforts! ❤️ 💛 💙 💚 💜
